File: Problems/Dynamic_Programming/12865.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <stdio.h>
#include <utility>
#include <cmath>

#define MAXN 101
#define MAXK 100001

using namespace std;

int n, k;

pair<int, int> knapsack[MAXN];
int D[MAXN][MAXK];

int solution(int index, int weight)
{
	if (index > n)
	{
		return 0;
	}
	
	int& result = D[index][weight];
	if (result)
	{
		return result;
	}
	result = solution(index + 1, weight);
	if (weight + knapsack[index].first <= k)
	{
		result = ::max(result, solution(index + 1, weight + knapsack[index].first) + knapsack[index].second);
	}

	return result;
}

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	//freopen("input.txt", "r", stdin);

	cin >> n >> k;
	for (int i = 1; i <= n; i++)
	{
		cin >> knapsack[i].first >> knapsack[i].second;
	}

	cout << solution(1, 0);

	return 0;
}

File: Krafton-Jungle/Week-04/12865.py
import sys


def solution(index: int, weight: int) -> int:
    global n, k
    if index == n:
        return 0
    if memo[index][weight]:
        return memo[index][weight]

    if knapsack[index][0] + weight > k:
        memo[index][weight] = solution(index + 1, weight)
    else:
        memo[index][weight] = max(
            solution(index + 1, weight),
            solution(index + 1, weight + knapsack[index][0]) + knapsack[index][1],
        )

    return memo[index][weight]


n, k = map(int, sys.stdin.readline().split())
knapsack = []
memo = [[0 for _ in range(k + 1)] for _ in range(n)]
for _ in range(n):
    knapsack.append(list(map(int, sys.stdin.readline().split())))

print(solution(0, 0))
